File: app/src/main/java/com/sjn/taggingplayer/db/dao/SongHistoryDao.java
package com.sjn.taggingplayer.db.dao;

import com.sjn.taggingplayer.db.SongHistory;

import java.util.Date;
import java.util.List;

import io.realm.Realm;
import io.realm.RealmQuery;
import io.realm.Sort;

public class SongHistoryDao extends BaseDao {

    private static SongHistoryDao sInstance;

    public static SongHistoryDao getInstance() {
        if (sInstance == null) {
            sInstance = new SongHistoryDao();
        }
        return sInstance;
    }

    public List<SongHistory> timeline(Realm realm, String recordType) {
        return realm.where(SongHistory.class).equalTo("mRecordType", recordType).findAllSorted("mRecordedAt", Sort.DESCENDING);
    }

    public List<SongHistory> where(Realm realm, Date from, Date to, String recordType) {
        RealmQuery<SongHistory> query = realm.where(SongHistory.class).equalTo("mRecordType", recordType);
        if (from != null) {
            query.greaterThanOrEqualTo("mRecordedAt", from);
        }
        if (to != null) {
            query.lessThanOrEqualTo("mRecordedAt", to);
        }
        return query.findAll();
    }

    public void save(Realm realm, final SongHistory songHistory) {
        realm.executeTransaction(new Realm.Transaction() {
            @Override
            public void execute(Realm realm) {
                songHistory.setId(getAutoIncrementId(realm, SongHistory.class));
                songHistory.setDevice(DeviceDao.getInstance().findOrCreate(realm, songHistory.getDevice()));
                songHistory.setSong(SongDao.getInstance().findOrCreate(realm, songHistory.getSong()));
                realm.insert(songHistory);
            }
        });
    }

    public void remove(Realm realm, final long id) {
        realm.executeTransactionAsync(new Realm.Transaction() {
            @Override
            public void execute(Realm realm) {
                realm.where(SongHistory.class).equalTo("mId", id).findFirst().deleteFromRealm();
            }
        });
    }

    public List<SongHistory> findAll(Realm realm) {
        return realm.where(SongHistory.class).findAll();
    }

    public SongHistory newStandalone() {
        return new SongHistory();
    }
}

File: app/src/main/java/com/sjn/stamp/model/dao/SongHistoryDao.kt
package com.sjn.stamp.model.dao

import com.sjn.stamp.model.Song
import com.sjn.stamp.model.SongHistory
import com.sjn.stamp.model.constant.RecordType
import io.realm.Realm
import io.realm.Sort
import java.util.*

object SongHistoryDao : BaseDao<SongHistory>() {

    fun where(realm: Realm, from: Date?, to: Date?, recordType: String): List<SongHistory> {
        val query = realm.where(SongHistory::class.java).equalTo("recordType", recordType)
        if (from != null) {
            query.greaterThanOrEqualTo("recordedAt", from)
        }
        if (to != null) {
            query.lessThanOrEqualTo("recordedAt", to)
        }
        return query.findAll() ?: emptyList()
    }

    fun timeline(realm: Realm, recordType: String): List<SongHistory> =
            findAll(realm, recordType)

    fun findAll(realm: Realm, recordType: String): List<SongHistory> =
            realm.where(SongHistory::class.java).equalTo("recordType", recordType).findAllSorted("recordedAt", Sort.DESCENDING) ?: emptyList()

    fun findPlayRecordByArtist(realm: Realm, artistName: String): List<SongHistory> =
            realm.where(SongHistory::class.java).equalTo("recordType", RecordType.PLAY.databaseValue).equalTo("song.artist.name", artistName).findAll() ?: emptyList()

    fun findOldestByArtist(realm: Realm, artistName: String): SongHistory? =
            realm.where(SongHistory::class.java).equalTo("recordType", RecordType.PLAY.databaseValue).equalTo("song.artist.name", artistName).findAllSorted("recordedAt", Sort.ASCENDING).first()

    fun findOldest(realm: Realm, songId: Long): SongHistory? =
            realm.where(SongHistory::class.java).equalTo("song.id", songId).equalTo("recordType", RecordType.PLAY.databaseValue).findAllSorted("recordedAt", Sort.ASCENDING).first()

    fun create(realm: Realm, rawSong: Song, recordType: RecordType, date: Date, count: Int) {
        SongDao.findById(realm, rawSong.id)?.let { song ->
            realm.executeTransaction({
                realm.copyToRealm(SongHistory().apply {
                    this.id = getAutoIncrementId(realm)
                    this.device = DeviceDao.findOrCreate(realm)
                    this.recordType = recordType.databaseValue
                    this.recordedAt = date
                    this.count = count
                    this.song = song
                })
            })
        }
    }

    fun delete(realm: Realm, songHistoryId: Long) {
        realm.executeTransactionAsync { r -> r.where(SongHistory::class.java).equalTo("id", songHistoryId).findFirst().deleteFromRealm() }
    }

}